### ISSUE TITLE
Clarify values for `code.function.name` and `code.namespace`

### ISSUE THREAD

**Issue by SylvainJuge** [open-telemetry/semantic-conventions#1677]
This is part of #1599 where we aim to make `code.*` attributes as release-candidate.

Some of the `code.*` attributes are being renamed with #1624, but this issue is about the values of those attributes and having a clear definition for them, in particular:

- `code.function.name` (previously `code.function`).
- `code.namespace`

In the discussion of #1624 we found that not having explicit per-language examples leaves interpretation open ([here](https://github.com/open-telemetry/semantic-conventions/pull/1624#discussion_r1870985463) and [here](https://github.com/open-telemetry/semantic-conventions/pull/1624#issuecomment-2534949287)), and is likely to cause ambiguity and potential inconsistencies, also there might be per-language constraints or overhead to provide those values.

Those attributes are currently in `experimental`, and we aim to promote them to `release candidate` with #1599, as per [this comment](https://github.com/open-telemetry/semantic-conventions/pull/1624#discussion_r1881292830) we consider that they are not used enough to justify a migration plan to prevent unexpected breaking changes (for example with `OTEL_SEMCONV_STABILITY_OPT_IN` environment variable).

Here the goal would be to define for each language:
- the value of `code.function.name`
- the value of `code.namespace`
- what value to use when only a single value is provided by the language/platform, in other words when to split and how
- what value to use when only a partial value is available, for example anonymous lambdas/functions that might not have an explicit name in code (but likely have a "technical name").

For now, we aim to maximize consistency, however if the overhead to provide those values (for example due to extra allocation or string splitting overhead), it might be possible to provide "slightly inconsistent values" to avoid those.

## Checklist
- [ ] Java [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2569081934) 
- [ ] Go [comment](https://github.com/open-telemetry/semantic-conventions/pull/1624#issuecomment-2534949287) for lambda, single value would be preferable but [could be split with minimal overhead](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/6415).
- [ ] PHP [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2576776269) and [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2577545831). single value would be preferable/simpler but could be split with minimal overhead.
- [ ] Python [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2575581985)
- [ ] Node/Javascript [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2576797137)
- [ ] Erlang/Elixir [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2576792921)
- [ ] Rust [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2585650950)
- [ ] Profiling [comment](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2590986638)
- [ ] ...

**Comment by pellared** [comment]
Wouldn't it be better/simpler to just have one attribute `code.function.name` which value is a fully-qualified name and get rid of `code.namespace`? It would solve both issues:

> - what value to use when only a single value is provided by the language/platform, in other words when to split and how
> - what value to use when only a partial value is available, for example anonymous lambdas/functions that might not have an explicit name in code (but likely have a "technical name").

Related comment: https://github.com/open-telemetry/semantic-conventions/pull/1624#issuecomment-2534949287

**Comment by SylvainJuge** [comment]
Having a "do it all" attribute might make sense, but I think that we need to first gather what would work best for each platform as separate fields, then we might merge them if that's a better approach. Here there might not be a single "best solution" as the values are platform/language dependent so choosing any is always a compromise.

**Comment by SylvainJuge** [comment]
For Java, what I think are the expected values for those attributes is the following as captured with [this gist](https://gist.github.com/SylvainJuge/62d66fa71af017661075d57b0080b24f) using the reflection API.

```
-- regular class
code.namespace = com.mycompany.MyClass
code.function.name = myMethod

-- anonymous class
code.namespace = com.mycompany.Main$1
code.function.name = myMethod

-- primitive type
code.namespace = int
code.function.name = n/a, not available for primitive types

-- lambda
code.namespace = com.mycompany.Main$$Lambda/0x0000748ae4149c00
code.function.name = myMethod
```

When using the reflection API, the values for `code.namespace` and `code.function.name` are each provided through a single method call, so no extra overhead nor processing is required.

Within bytebuddy advices, we can also get those through `@Advice.Origin` annotation, either as separate values or combined in a single one. When outside of instrumentation advices, for example with "inferred spans" which are generated by a sampling profiler in Java, an equivalent name will have to be captured and thus might require some minimal string processing.

So in the case of Java, I think that having separate attributes is probably the best option.

**Comment by trask** [comment]
@open-telemetry/dotnet-approvers
@open-telemetry/cpp-approvers 
@open-telemetry/erlang-approvers 
@open-telemetry/go-approvers 
@open-telemetry/javascript-approvers 
@open-telemetry/php-approvers 
@open-telemetry/rust-approvers 
@open-telemetry/swift-approvers 

could you help us out and post a common example(s) of what would be most common/expected to capture in your language for `code.namespace` and `code.function`? thanks!

**Comment by brettmc** [comment]
For PHP, we extensive use these for auto-instrumentation of a function or method call. For methods, `code.namespace` is the FQN of the class that the method belongs to. It may be blank for a global or built-in function.
`code.function` represents the function (or method) that was instrumented.

From one of our tests, `GuzzleHttp\Client::transfer()` was instrumented here:
```
["code.function"]=> string(8) "transfer"
["code.namespace"]=> string(17) "GuzzleHttp\Client"
```

I don't see any issue with us merging these into a single field.

**Comment by bryannaegele** [comment]
Erlang/Elixir would be fully qualified module name for namespace and function/arity for function name.

**Comment by pellared** [comment]
Go:

```
-- regular function
code.namespace = github.com/my/repo/pkg
code.function.name = foo

-- anonymous function (inside foo function)
code.namespace = github.com/my/repo/pkg.foo
code.function.name = func5 // or other funcN generated by the compiler where N is a positive integer
```

I it worth mentioning that the Go runtime operates on fully qualified names (e.g. `github.com/my/repo/pkg.foo.func5`) and we have to manually spit the value:

```go
// splitFuncName splits package path-qualified function name into
// function name and package full name (namespace). E.g. it splits
// "github.com/my/repo/pkg.foo" into
// "foo" and "github.com/my/repo/pkg".
func splitFuncName(f string) (string, string) {
	i := strings.LastIndexByte(f, '.')
	if i < 0 {
		return "", ""
	}
	return f[i+1:], f[:i]
}
```

**Comment by intuibase** [comment]
> For PHP, we extensive use these for auto-instrumentation of a function or method call. For methods, `code.namespace` is the FQN of the class that the method belongs to. It may be blank for a global or built-in function. `code.function` represents the function (or method) that was instrumented.
> 
> From one of our tests, `GuzzleHttp\Client::transfer()` was instrumented here:
> 
> ```
> ["code.function"]=> string(8) "transfer"
> ["code.namespace"]=> string(17) "GuzzleHttp\Client"
> ```

I would like to add that for PHP there might be few cases:

For a function defined in an anonymous namespace:
```code.function: FunctionName```

For a function defined in a named namespace:
```code.function: Namespace\FunctionName```

In this case, code.namespace will always be missing. In my opinion, this reveals an inconsistency in the interpretation of what constitutes a "namespace" versus a "function". This is due to the way PHP stores function names - introducing a split would create additional overhead.

For a class method defined in an anonymous namespace:
```
code.namespace: ClassName
code.function: MerhodName
```

For a class method defined in a named namespace:
```
code.namespace: Namespace\ClassName
code.function: MerhodName
```

**Comment by SylvainJuge** [comment]
@intuibase do we have an idea of the overhead that splitting would cause here ? For Go the conclusion was that it is negligible, but not zero. The original intent here is to favor consistency, but if the overhead becomes too important then we could keep some per-platform inconsistencies and document them.

**Comment by hdost** [comment]
Using this playground as an example https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3196b0d9d24d31c3dafa15cb6769e9f1

The backtrace looks something like this:
```
playground::my_module::my_cool_func
             at /playground/src/main.rs:8:18
...
```

So the value should map like this:
```
code.function.name: my_cool_func
code.namespace: playground::my_module
```

The context is that crates have a name which is included. 

Reference: https://doc.rust-lang.org/reference/names/namespaces.html

**Comment by trask** [comment]
interestingly, the profiling signal appears to be modeling "namespace + function" as a single field:

https://github.com/open-telemetry/opentelemetry-proto/blob/ffade295895a2be5a6e7931eb0cda1c72bc4c0f6/opentelemetry/proto/profiles/v1development/profiles.proto#L467

there could be benefit in alignment

it may help to also consider if there's alignment to be had with (future) https://github.com/open-telemetry/opentelemetry-specification/issues/2839

**Comment by trask** [comment]
> For the profiling signal, do we have an available specification of the content of those attributes?

@open-telemetry/profiling-approvers do you have any spec or examples for what is put into the [`Function.name`](https://github.com/open-telemetry/opentelemetry-proto/blob/ffade295895a2be5a6e7931eb0cda1c72bc4c0f6/opentelemetry/proto/profiles/v1development/profiles.proto#L467) proto field for different languages? thanks!

**Comment by SylvainJuge** [comment]
Thanks @christos68k, this pushes a bit more on having a single attribute for the function name.

For now, this single "function name" field is in the profiling data model (defined in protobuf), and the value really depends on the implementation of the unwinder/symbolizer, which brings at least two ideas/questions:

First, aligning perfectly on the values in profiling and semconv for every platform/language might be desirable, but hard to achieve. In particular if we take the constraints and implementation details of the unwinder/symbolizer having to normalize and/or split fields might be tricky. However, doing some post-processing where this profiling data is produced seems possible.

Then, the profiling data model expressed in the protobuf is mostly a "wire protocol" in the sense that it reflects how the data is being captured and stored efficiently, and as such it would be hard to add extra constraints from the semantic conventions on how things should be structured. What I think we need to provide is the ability to correlate the values in "profiling protobuf" and "code.* semantic conventions", even if that means 1 field in profiling would be split in multiple ones in semconv.

If we take the example of the JVM, we can see that the profiling function name contains also the arguments and return value types, and currently we don't have any semconv attributes to store them (as a complete method signature or as individual attributes), so the best we could do for now is a partial match between semconv attributes and profiling data.

**Comment by pellared** [comment]
> this pushes a bit more on having a single attribute for the function name

I never studied compilers/interpreters but maybe there is some better name like `code.symbol.name`. I think that the term function is language-specific (e.g. some languages have methods).  Reference: https://en.wikipedia.org/wiki/Executable_and_Linkable_Format

**Comment by SylvainJuge** [comment]
After talking a bit with @christos68k today, I can confirm the following:
- the profiling protobuf is a "wire protocol" that originates from [pprof](https://github.com/google/pprof/tree/main) and is thus optimized for transport/compacity, thus using semconv attributes in profiling data would cause huge overhead and is not an option.
- there are only a few [process semantic convention attributes](https://opentelemetry.io/docs/specs/semconv/attributes-registry/process/#process-attributes) that are [mapped to profiling data model](https://opentelemetry.io/docs/specs/otel/profiles/mappings/)

In addition, as we can see in the [provided examples here](https://github.com/open-telemetry/semantic-conventions/issues/1677#issuecomment-2590986638), the values that we could map to `code.*` semconv attributes are not structured and their interpretation will be per language and runtime environment. Also, the symbolization process that allows to get those "human friendly names" is something that is done using background processing at the collector level and relies on sometimes complex heuristics to resolve those.

Also, the profiling data described here is only the "raw data" that needs to be aggregated/sampled/processed on the backend. Due to the volume of profiling data, the backend storage will always need to be tailor-made for the use-case, this part of the mapping to semantic conventions will always be optional (for example if you create a product that only aggregates at process level, then those code-level details might not require to be mapped to semconv).

With that in mind, this means for profiling data:
- we can't expect the profiling data `Function.name` to be properly structured nor easily specified
- the mapping between `Function.name` and any potential semconv attribute(s) will be loose and it's not possible to correlate 1:1 with a single attribute in semconv.

If there is any mapping between `Function.name` and semconv attribute(s), we should do it in a "best effort" and that will require some per-language heuristics. Also, this should be a recommendation rather than a strict requirement because we expect things to not align 100% of the time.

----

Going back to the main issue discussion, so far I think we should move forward with the proposal to have `code.namespace` and `code.function.name` for the following reasons:
- the terms are generic enough to fit most languages, if there are any ambiguities we can lift them with examples
- while there could be some overhead to split, it seems mostly negligible most of the time
- if there are per-language exceptions, we can properly document them

**Comment by trask** [comment]
@SylvainJuge I think a good next step would be to write up what should be captured in the `code.*` attributes for as many of the above languages as you can, and send a PR with that to https://github.com/open-telemetry/semantic-conventions/tree/main/docs/code/README.md

**Comment by pellared** [comment]
What are the benefits of having two attributes instead of one? The issue mostly describes the issues connected with the split but I do not see any description of benefits of moving  forward with this approach.

When reading through this issue I see mostly reasons to use a single attribute.

**Comment by trask** [comment]
> When reading through this issue I see mostly reasons to use a single attribute.

@pellared could you summarize these reasons for us?

**Comment by pellared** [comment]
1. Ambiguity when a language needs to split some value when it's API returns FQN.
2. Profilers and stacktraces usually report FQN
3. Performance overhead caused by a split
4. Unclear benefits of having the attributes separated. Having FQN in one attribute seems more compelte and self-contained. I do not see any use case when having just "function name" (without the namespace) would be useful.

**Comment by SylvainJuge** [comment]
In addition to the reasons provided by @pellared in favor of a "single field" approach, we can also add the following:
- splitting with a per-language separator between namespace and function, for example `.` or `#` in Java, or `::` in PHP and Rust means the captured string will likely have to be reconstructed later with the same separator to provide a value familiar to the end-user.

So, I agree that it would be simpler only a single attribute and I'll try to open a PR to document this approach.

If we deprecate the `code.namespace` attribute, can we still get away with not providing a safe opt-in with `OTEL_SEMCONV_STABILITY_OPT_IN` or equivalent ?

However, given the attribute `code.function.name` explicitly refers to the function name, I think that everything that is not the function name would still need to be removed, in the examples that were provided so far we have:
- Erlang/Elixir: function arity would have to be removed
- For sampling profiler JVM frames: method return type and arguments would have to be removed

Also, doing so would allow to have a more "strict" definition and could allow to have a 1:1 matching between profiler frames and the `code.function.name` attribute with the help of some post-processing of profiling data.

In both cases those elements participate to define the unique function identity (for example with JVM you can have identical methods that only differ by return type), however they might not be always available, so I would suggest to cover those if needed later with a separate attribute like `code.function.arguments` and `code.function.returntype`.

Another benefit of this approach I see is that it's easier to add extra attributes once `code.function.name` is stable rather than deprecating it and splitting because we later decide not to include those args/return value in it.

In summary I'd like to have feedback on the following aspects:
- If we deprecate `code.namespace`, do we have to provide opt-in option ? Or could we simply leave `code.namespace` as experimental while other `code.*` attributes are marked as stable and deal with removal later ?
- Do you agree that removing arguments/return values from the value is a good compromise for now and we can improve on this later ?

**Comment by trask** [comment]
> If we deprecate `code.namespace`, do we have to provide opt-in option ? Or could we simply leave `code.namespace` as experimental while other `code.*` attributes are marked as stable and deal with removal later ?

I don't think we have to provide a backcompat option and instrumentations can decide how to handle upgrading, since the decision will be driven by usage. For example, in Java instrumentation since we populate `code.*` on the fairly heavily used `@WithSpan`, we probably will want to wait and make this change in a 3.0 major version bump, but this may not be needed in other languages.

> * Do you agree that removing arguments/return values from the value is a good compromise for now and we can improve on this later ?

👍
